pennylane/templates/subroutines/qrom.py

```diff
@@ -0,0 +1,299 @@
+
+
+        The ``work_wires`` are the auxiliary qubits used by the template to reduce the number of gates required.
+        Let :math:`k` be the number of work wires. If :math:`k = 0`, the template is equivalent to executing :class:`~.Select`.
+        Following the idea in `arXiv:1812.00954 <https://arxiv.org/abs/1812.00954>`__, auxiliary qubits can be used to
+        load more than one bitstring in parallel . Let :math:`\lambda` be
+        the number of bitstrings we want to store in parallel, assumed to be a power of :math:`2`.
+        Then, :math:`k = l \cdot (\lambda-1)` work wires are needed,
+        where :math:`l` is the length of the bitstrings.
+
+        The QROM template has two variants. The first one (``clean = False``) is based on `arXiv:1812.00954 <https://arxiv.org/abs/1812.00954>`__ that alterates the state in the ``work_wires``.
+        The second one (``clean = True``), based on `arXiv:1902.02134 <https://arxiv.org/abs/1902.02134>`__, solves that issue by
+        returning ``work_wires`` to their initial state. This technique can be applied when the ``work_wires`` are not
+        initialized to zero.
+
+    """
+
+    def __init__(
+        self, bitstrings, control_wires, target_wires, work_wires, clean=True, id=None
+    ):  # pylint: disable=too-many-arguments
+
+        control_wires = qml.wires.Wires(control_wires)
+        target_wires = qml.wires.Wires(target_wires)
+
+        work_wires = qml.wires.Wires(work_wires) if work_wires else qml.wires.Wires([])
+
+        self.hyperparameters["bitstrings"] = bitstrings
+        self.hyperparameters["control_wires"] = control_wires
+        self.hyperparameters["target_wires"] = target_wires
+        self.hyperparameters["work_wires"] = work_wires
+        self.hyperparameters["clean"] = clean
+
+        if work_wires:
+            if any(wire in work_wires for wire in control_wires):
+                raise ValueError("Control wires should be different from work wires.")
+
+            if any(wire in work_wires for wire in target_wires):
+                raise ValueError("Target wires should be different from work wires.")
+
+        if any(wire in control_wires for wire in target_wires):
+            raise ValueError("Target wires should be different from control wires.")
+
+        if 2 ** len(control_wires) < len(bitstrings):
+            raise ValueError(
+                f"Not enough control wires ({len(control_wires)}) for the desired number of "
+                + f"bitstrings ({len(bitstrings)}). At least {int(math.ceil(math.log2(len(bitstrings))))} control "
+                + "wires required."
+            )
+
+        if len(bitstrings[0]) != len(target_wires):
+            raise ValueError("Bitstring length must match the number of target wires.")
+
+        all_wires = target_wires + control_wires + work_wires
+        super().__init__(wires=all_wires, id=id)
+
+    def _flatten(self):
+        data = (self.hyperparameters["bitstrings"],)
```

Comment:
Since `bitstrings` contains strings, it is not traceable data, so it should be in `metadata` instead.  We just need to make sure that it always is a hashable tuple instead of being a list.